Services/CO.CDP.Organisation.WebApi/UseCase/LookupOrganisationUseCase.cs

```diff
@@ -3,11 +3,58 @@
 using CO.CDP.OrganisationInformation.Persistence;
 
 namespace CO.CDP.Organisation.WebApi.UseCase;
-public class LookupOrganisationUseCase(IOrganisationRepository organisationRepository, IMapper mapper) : IUseCase<string, Model.Organisation?>
+public class LookupOrganisationUseCase(IOrganisationRepository organisationRepository, IMapper mapper) : IUseCase<OrganisationQuery, Model.Organisation?>
 {
-    public async Task<Model.Organisation?> Execute(string name)
+    public async Task<Model.Organisation?> Execute(OrganisationQuery query)
     {
-        return await organisationRepository.FindByName(name)
-            .AndThen(mapper.Map<Model.Organisation>);
+        if (!query.IsValid)
+        {
+            throw new ArgumentException("Organisation name or identifier must be provided.");
+        }
+
+        if (!string.IsNullOrEmpty(query.Identifier))
+        {
+            if (IsIdentifier(query.Identifier, out var scheme, out var id))
+            {
+                return await organisationRepository.FindByIdentifier(scheme, id)
+                    .AndThen(mapper.Map<Model.Organisation>);
+            }
+            throw new ArgumentException("Invalid identifier format.");
+        }
+
+        if (!string.IsNullOrEmpty(query.Name))
+        {
+            return await organisationRepository.FindByName(query.Name)
+                .AndThen(mapper.Map<Model.Organisation>);
+        }
+
+        return null;
     }
+    private bool IsIdentifier(string query, out string scheme, out string id)
+    {
+        var parts = query.Split(':');
+        if (parts.Length == 2)
+        {
+            scheme = parts[0];
+            id = parts[1];
+            return true;
+        }
+        scheme = string.Empty;
+        id = string.Empty;
+        return false;
+    }
+}
+
+public class OrganisationQuery
```

Comment:
As this is a value object, a record would be a better choice.

Reply:
Hi Kuba, Updated this.